### PR TITLE
Editable annotations and persistent selection overlays

### LINKED ISSUE #3412: Add support for persistent selection overlays
Currently, when you select an area with the lasso tool, and release the mouse button, the dotted line area disappears. An option to keep the lasso outline visible might be a nice feature.

### LINKED ISSUE #5829: PolySelectTool area remains the same at all zoom levels
When trying to select a series of points using the PolySelectTool, the area selected doesn't respect the zoom level of the plot. Particularly for the mapping application I'm working on, this is extremely undesirable behavior, as it precludes precise selection of points.

I first mentioned this over on the Google Group [here](https://groups.google.com/a/continuum.io/forum/#!topic/bokeh/m3JzMUFub9w). I was asked to open an issue here for the bug. Example screenshots from that post: 
image
image

Here's the minimal code to reproduce the bug (area/points displayed differ from screenshots above). The two lines can be uncommented to show the issue on a tile map.

```python
from bokeh.io import (
    show,
    output_file
)
from bokeh.models import (
    PolySelectTool,
    ColumnDataSource,
    Circle,
    WheelZoomTool,
    PanTool
)
from bokeh.plotting import figure
#from bokeh.tile_providers import CARTODBPOSITRON
source = ColumnDataSource(
    {'coordsX':[-8215213.30496,
                -8173542.15792,
                -8207605.64239],
     'coordsY':[5265013.23126,
                5198976.24299,
                5267118.16664]})
p = figure(tools=["wheel_zoom","pan","poly_select"])
p.circle(x='coordsX',y='coordsY',source=source)
#p.add_tile(CARTODBPOSITRON)
output_file('poly-bug.html')
show(p)
```

Running Bokeh 0.12.4 on Linux Mint / Ubuntu inside a virtualenv. Let me know if I need to provide further information.

### LINKED ISSUE #8180: HoverTool not working with RangeTool
On Bokeh 0.13.0 and Firefox 61.0.1 the HoverTool won't show the tooltip on the areas covered by the BoxAnnotation of the RangeTool. I would expect the HoverTool to work regardless of where the RangeTool is located.

Please find some code reproducing the issue below.

```
from bokeh.layouts import row
from bokeh.models import RangeTool, ColumnDataSource, HoverTool, Range1d
from bokeh.plotting import  output_file, show, figure
from numpy import linspace, cos, pi

x = linspace(0, 4)
y = cos(2*pi*x)

plot = figure()
plot.line('x', 'y', source=ColumnDataSource(data=dict(x=x, y=y)))


range_tool = RangeTool(x_range=Range1d(1, 2))
range_tool.overlay.fill_alpha = 0.2
plot.add_tools(range_tool)

plot.add_tools(HoverTool(
    tooltips=[
        ("(x,y)", "(@x, @y)"),
    ],
    mode='vline'
))

output_file('hover_plus_range.html')
show(row(plot))

```

### LINKED ISSUE #9047: [BUG] Can't interact with multiple RangeTools
#### ALL software version info (bokeh, python, notebook, OS, browser, any other relevant packages)

bokeh.version: 1.2.0
Python version: 3.5.3
OS: Debian GNU/Linux 9.9 (stretch) 
System info: SMP Debian 4.9.144-3.1 (2019-02-19) x86_64
Browser: Google Chrome 75.0.3770.100 and Mozilla Firefox 60.7.2

#### Description of expected behavior and the observed behavior

I'm working on a customised version of the [ridge plot](https://bokeh.pydata.org/en/latest/docs/gallery/ridgeplot.html) with multiple patches. What I wanted was a `RangeTool` object _for each patch of my plot_.  I'm currently using the `add_tools()` method from `plotting.figure` inside a for loop, but even though those multiple `RangeTool` objects do appear as expected, I can only interact with the first one, which means I'm not able to drag or stretch the remaining.

#### Complete, minimal, self-contained example code that reproduces the issue

```python
from bokeh.models import ColumnDataSource, RangeTool, Range1d
from bokeh.plotting import figure
from bokeh.io import show

import numpy as np

NPOINTS = 100
labels = ['Time','To','Be','Free']

def ridge(category, data=None, scale=1.0):
    return list(zip([category]*len(data), scale*data))

p = figure(title='Horizon chart wannabe',
        y_range=labels + [''], # FIXME y_max padd workaround
        x_range=(0, NPOINTS),
        plot_width=1000, plot_height=600,
        toolbar_location=None)

for i, lab in enumerate(labels):
    source = ColumnDataSource(data=dict(x=np.arange(NPOINTS),))
    vector = np.random.rand(NPOINTS-2)
    vector = np.insert(vector, 0, 0) # FIXME
    vector = np.append(vector, 0)    # FIXME 
    y = ridge(lab, data=vector)
    source.add(y, lab)

    renderer = p.patch('x', lab, name=lab,
            color='red', line_color='red', 
            fill_alpha=0.4, line_alpha=0.1,
            source=source)

    range_tool = RangeTool(
            x_range=Range1d((i+1)*5, (i+1)*5+5),
            y_range=Range1d(i+0.5, i+1.5), y_interaction=False)
    range_tool.overlay.fill_color = "navy"
    range_tool.overlay.fill_alpha = 0.2

    p.add_tools(range_tool) # TODO add to `renderer` patch?

show(p)
```

#### Screenshots or screencasts of the bug in action
The red circle comes from the screen recorder and appears when I click with the mouse. It is not related to Bokeh.
bug

### LINKED ISSUE #9381: [BUG] range_tool broken on inverted axes
#### ALL software version info (bokeh, python, notebook, OS, browser, any other relevant packages)

bokeh.version: 1.3.4
Python version: 3.7.3
OS: GNU/Linux x86_64
Browser: Google Chrome 77.0.3865.90
JupyterLab version: 1.1.3

#### Description of expected behavior and the observed behavior

When a RangeTool is used with an inverted axis (i.e. a y axis in which the data increase downwards), the range tool is unresponsive.

In this example, the RangeTool should allow a user to control the zoom of the figure on the right, but the shaded area on the left cannot be moved up or down.

However, when a user pans the plot on the right, the RangeTool does still move up and down as expected.

The bug seems to only appear when the Range1d is inverted.

#### Complete, minimal, self-contained example code that reproduces the issue

```
from bokeh.models.ranges import Range1d
from bokeh.layouts import row
from bokeh.models import RangeTool
from bokeh.plotting import figure, show

xx = [20,22,21]
yy = [2,3,4]
kwgs = dict(plot_height=300, plot_width=300)

overview_rng = Range1d(5, 0)   # Range is inverted, higher number first
zoomed_rng = Range1d(4, 2)

fig_overview = figure(y_range=overview_rng, **kwgs)
fig_overview.line(xx, yy)

range_tool = RangeTool(y_range=zoomed_rng)
fig_overview.add_tools(range_tool)

fig_zoomed = figure(y_range=zoomed_rng, **kwgs)
fig_zoomed.line(xx, yy)

show(row([fig_overview, fig_zoomed]))
```

#### Screenshots or screencasts of the bug in action

range_tool_inverted

### LINKED ISSUE #12731: [BUG] Possible regression: pinching not working with wheelzoomtool
### Software versions

Python version      :  3.9.9 (main, Jan  1 1970, 00:00:01) 
IPython version     :  8.2.0
Tornado version     :  5.1.1
Bokeh version       :  3.0.3
BokehJS static path :  /home/eris/.local/lib/python3.9/site-packages/bokeh/server/static
node.js version     :  (not installed)
npm version         :  (not installed)
Operating system    :  Linux-6.0.16-x86_64-with-glibc2.33

### Browser name and version

Different behaviors in Firefox (108.0.1 (64-bit)) and Webkit (WebKitGTK 2.36.8)

### Jupyter notebook / Jupyter Lab version

_No response_

### Expected behavior

Active Wheelzoom should enable zoom on pinch and avoid the browsers default action of zooming the viewport.

In other words: zooming YES, viewport zoomed NO.

### Observed behavior

On Firefox (108.0.1 (64-bit)): zooming NO, viewport zoomed YES.

On Webkit (WebKitGTK 2.36.8): zooming YES, viewport zoomed YES.

### Example code

```Python
from bokeh.models import ColumnDataSource
from bokeh.plotting import figure, output_file, save

cds = ColumnDataSource(
    dict(
        x=[0],
        y=[0],
        w=[1],
        h=[1],
    )
)
fig = figure(
    x_range=(-2, 2),
    y_range=(-2, 2),
    width=800,
    height=600,
    tools=['ywheel_zoom'],
    active_scroll="ywheel_zoom",
)
fig.rect(
    x="x",
    y="y",
    width="w",
    height="h",
    source=cds,
)
output_file("test_pinch_screen.html", mode="inline")
save(fig)
```

### Stack traceback or browser console output

_No response_

### Screenshots

_No response_

### PR DESCRIPTION
**Pull request by mattpap**
Extracted relevant changes from PR #12074.

TODO:

- [x] investigate not working modes with `select_every_mousemove`
- [x] limit editability when `dimension != "both"`
- [x] persist the overlay when switching tools (allow pan, zoom, etc.)
- [x] better editability for `origin == "center"`
- [x] redesign `RangeTool` using editable annotations
- [x] add support for move/resize limits
- [x] allow event propagation (e.g. hover passthrough)
- [x] allow alt key to override `select_every_mousemove`
- [x] resolve corner cases of overlay persistence across tool switching
- [x] add interactive visual integration tests

These are partially done in separate work:

- [ ] formalize multiple selection overlay support
- [ ] add support for resize/move handles

Example (from #12074, but it works the same here):

https://user-images.githubusercontent.com/27475/161348036-b63cb3c2-9b1d-426d-ab9e-6cd58dd1a7c2.mp4

- [x] fixes #3412
- [x] fixes #5829
- [x] fixes #8180
- [x] fixes #9047
- [x] fixes #9381
- [x] fixes #12731

### CODE REVIEW

File: bokehjs/src/lib/models/annotations/box_annotation.ts [hunk 1]
```diff
@@ -72,24 +87,160 @@ export class BoxAnnotationView extends AnnotationView {
...
-    else if (Math.abs(sy-bottom) < tol || Math.abs(sy-top) < tol)
-      return this.model.ns_cursor
-    else if (this.bbox.contains(sx, sy))
-      return this.model.in_cursor
-    else
-      return null
+    const tolerance = Math.max(EDGE_TOLERANCE, this.model.line_width/2)
+
+    const hits_left = abs(left - sx) < tolerance
+    const hits_right = abs(right - sx) < tolerance
```
Comment (ianthomas23) [inline bokehjs/src/lib/models/annotations/box_annotation.ts:101]: If the box is really narrow, this can result in both `hits_left` and `hits_right` being `true` and subsequently the left is preferred, even if the cursor is just to the right of both edges. I think we can do better by only allowing one of the two to be set, the edge that is nearest to the cursor position. This can be achieved by adding a subsequent check something like (untested):
```typescript
if (hits_left && hits_right) {
  if (abs(left - sx) < abs(right - sx))
    hits_right = false
  else
    hits_left = false
}
```
or alternatively finding the nearest edge first and only checking the tolerance for that edge.

And the same for `hits_top` and `hits_bottom` too.

**Comment by mattpap** [comment]
This is functionally ready. I'm still working on integration tests for the new features. The only new public non-experimental API is `RegionSelectTool.persistent`. All APIs related to "editability" of annotations are highly experimental and marked as such. The implementation of editable annotations is a subject for deep changes in the future. Originally this PR was much larger, but I extracted too far reaching changes to separate work, which I will be publishing for 3.2 (e.g. editable `Label`, movable/droppable `Legend` or `Axis`).

File: src/bokeh/models/annotations/geometry.py [hunk 2]
```diff
@@ -111,6 +113,44 @@ def __init__(self, *args, **kwargs) -> None:
     default.
     """)
 
+    border_radius = BorderRadius(default=0, help="""
+    Allows the box to have rounded corners.
+
+    .. note::
+        This property is experimental and may change at any point.
+    """)
```
Comment (bryevdv) [inline src/bokeh/models/annotations/geometry.py:121]: For my own near future reference, do you prefer

```python
border_radius = Experimental(BorderRadius(...))
```
or
```python
border_radius = BorderRadius(..., experimental=True)
```
or 
```python
border_radius = BorderRadius(...).experimental()
```
or some other kind of programmatic marking? 

I don't really like the first, since "experimental" is not a type, per se, but am open to whichever.

Comment (mattpap) [reply src/bokeh/models/annotations/geometry.py:121]: Definitely not the second one, as it "infects" APIs, and I've been working on removing this pattern from bokeh. Probably the first, which is similar to `Alias` and `DeprecatedAlias` and some other. If I was to implement this, I would choose the first one.

File: examples/interaction/tools/persistent_selections.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+          line_color=None)
+
+box_select = p.select_one(BoxSelectTool)
+box_select.persistent = True
+
+poly_select = p.select_one(PolySelectTool)
+poly_select.persistent = True
+
+lasso_select = p.select_one(LassoSelectTool)
+lasso_select.persistent = True
```
Comment (bryevdv) [inline examples/interaction/tools/persistent_selections.py]: Suggest this is preferable:
```suggestion
p.select_one(BoxSelectTool).persistent = True
p.select_one(PolySelectTool).persistent = True
p.select_one(LassoSelectTool).persistent = True
```

File: examples/interaction/tools/persistent_selections.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+from bokeh.models import BoxSelectTool, LassoSelectTool, PolySelectTool
+from bokeh.plotting import figure, show
+
+N = 2000
+x = np.random.random(size=N) * 100
+y = np.random.random(size=N) * 100
+radii = np.random.random(size=N) * 1.5
+colors = np.array([ [r, g, 150] for r, g in zip(50 + 2*x, 30 + 2*y) ], dtype="uint8")
+
+TOOLS="hover,crosshair,pan,wheel_zoom,zoom_in,zoom_out,box_zoom,undo,redo,reset,save,box_select,poly_select,lasso_select,help"
```
Comment (bryevdv) [inline examples/interaction/tools/persistent_selections.py]: can we remove some of these? Most of them are unrelated and just clutter the example

File: bokehjs/src/lib/core/ui_events.ts [hunk 5]
```diff
@@ -416,6 +484,84 @@ export class UIEventBus implements EventListenerObject {
     const base_type = event_type.split(":")[0] as BaseType
     const view = this._hit_test_renderers(plot_view, e.sx, e.sy)
 
+    if (base_type == "pan") {
+      if (this._current_pan_view == null) {
+        if (view != null) {
+          if (event_type == "pan:start" && is_Pannable(view)) {
+            if (view._pan_start(e as PanEvent)) {
```
Comment (ianthomas23) [inline bokehjs/src/lib/core/ui_events.ts:491]: Is there a particular reason to have these if-statements separate rather than combining them together?

File: src/bokeh/models/annotations/geometry.py [hunk 6]
```diff
@@ -111,6 +113,44 @@ def __init__(self, *args, **kwargs) -> None:
...
+    editable = Bool(default=False, help="""
+    Allows to interactively modify the geometry of this box.
+
+    .. note::
+        This property is experimental and may change at any point.
+    """)
+
+    resizable = Enum("none", "left", "right", "top", "bottom", "x", "y", "all", default="all", help="""
+    If `editable` is set, this property allows to configure which
+    combinations of edges are allwed to be moved, thus allows
```
Comment (ianthomas23) [inline src/bokeh/models/annotations/geometry.py]:
```suggestion
    combinations of edges are allowed to be moved, thus allows
```

File: src/bokeh/models/annotations/geometry.py [hunk 7]
```diff
@@ -218,13 +281,29 @@ def __init__(self, *args, **kwargs) -> None:
     The {prop} values for the polygon.
     """)
 
-    line_alpha = Override(default=0.3)
+    hover_line_props = Include(ScalarLineProps, prefix="hover", help="""
+    The {prop} values for the box when hovering over.
```
Comment (ianthomas23) [inline src/bokeh/models/annotations/geometry.py]:
```suggestion
    The {prop} values for the polygon when hovering over.
```

File: src/bokeh/models/annotations/geometry.py [hunk 8]
```diff
@@ -218,13 +281,29 @@ def __init__(self, *args, **kwargs) -> None:
...
+    """)
+
+    hover_fill_props = Include(ScalarFillProps, prefix="hover", help="""
+    The {prop} values for the box when hovering over.
```
Comment (ianthomas23) [inline src/bokeh/models/annotations/geometry.py]:
```suggestion
    The {prop} values for the polygon when hovering over.
```

File: src/bokeh/models/annotations/geometry.py [hunk 9]
```diff
@@ -218,13 +281,29 @@ def __init__(self, *args, **kwargs) -> None:
...
+    """)
+
+    hover_hatch_props = Include(ScalarHatchProps, prefix="hover", help="""
+    The {prop} values for the box when hovering over.
```
Comment (ianthomas23) [inline src/bokeh/models/annotations/geometry.py]:
```suggestion
    The {prop} values for the polygon when hovering over.
```

File: src/bokeh/models/annotations/geometry.py [hunk 10]
```diff
@@ -296,10 +375,24 @@ def __init__(self, *args, **kwargs) -> None:
...
+    .. note::
+        This property is experimental and may change at any point.
+    """)
+
     line_props = Include(ScalarLineProps, help="""
     The {prop} values for the span.
     """)
 
+    hover_line_props = Include(ScalarLineProps, prefix="hover", help="""
+    The {prop} values for the box when hovering over.
```
Comment (ianthomas23) [inline src/bokeh/models/annotations/geometry.py]:
```suggestion
    The {prop} values for the span when hovering over.
```

**Review by ianthomas23 — COMMENTED** [review]
I've tried out the new functionality here and it is very nice. I didn't know in advance how to cancel a persistent selection but I found out quite quickly by experimentation so I think that means it is intuitive.

One improvement from a usability point of view would be different cursors for editing nodes, editing edges and moving a poly annotation as at the moment they are the same so it is not possible to visually identify if you are on top of a node or edge before doing the edit. But I suppose we are limited by the standard set of cursors.

**Comment by mattpap** [comment]
> I've tried out the new functionality here and it is very nice. I didn't know in advance how to cancel a persistent selection but I found out quite quickly by experimentation so I think that means it is intuitive.

In near future (it was originally part of this PR, but I moved it out for now) there will be be possibility to interact with any canvas UI element with a context menu (at least on the desktop; mobile situation to be figured out), which in the case of persistent selection overlays would provide the option to clear the overlay and the selection. That should help with discoverability, because the user will be one click away from seeing entity specific options. Even more important if a toolbar is not visible, but tools are present nevertheless.

> One improvement from a usability point of view would be different cursors for editing nodes, editing edges and moving a poly annotation as at the moment they are the same so it is not possible to visually identify if you are on top of a node or edge before doing the edit. But I suppose we are limited by the standard set of cursors.

This was implemented a long time ago in PR #12083 and now I'm reimplementing this on top of this work. Cursors can be any image, in particular SVG images, so we have all the freedom to provide customized cursors for every application.

File: bokehjs/src/lib/core/ui_events.ts (hunk 5 above)
Comment (mattpap) [reply bokehjs/src/lib/core/ui_events.ts:491]: It's either a leftover from a more complex implementation or I may have more changes coming at some point. Probably a combination of both.

File: examples/interaction/tools/persistent_selections.py (hunk 3 above)
Comment (mattpap) [reply examples/interaction/tools/persistent_selections.py]: The reason this code is structured this way, is that this example originally showed how to configure tools' overlays, like restricting editability/movement, setting up handles, etc. Some of those options were postponed and all other are experimental, thus currently this example only sets `persistent`.

File: examples/interaction/tools/persistent_selections.py (hunk 4 above)
Comment (mattpap) [reply examples/interaction/tools/persistent_selections.py]: For the most part they aren't unrelated. For example, crosshair and hover tools are important to show how UI events are propagated. Pan and zoom tools are useful for showing how overlays are persisted between different interactions. I refactored the example to get rid of the list, but otherwise most of those tools should stay, maybe with some commentary what to look for.

Comment (bryevdv) [reply examples/interaction/tools/persistent_selections.py]: The updated version seems better, thanks

File: bokehjs/src/lib/core/ui_events.ts (hunk 5 above)
Comment (ianthomas23) [reply bokehjs/src/lib/core/ui_events.ts:491]: OK.

**Review by ianthomas23 — APPROVED** [review]
All my comments have been dealt with, so I am approving this.

There are some test failures.

**Comment by mattpap** [comment]
This is fully ready now. More testing and documentation will follow next week. Unless there are any objections, I would like to merged this soon.
